lib/Sema/CSDiagnostics.cpp

```diff
@@ -5490,12 +5491,93 @@ SourceLoc SkipUnhandledConstructInFunctionBuilderFailure::getLoc() const {
   return unhandled.get<Decl *>()->getLoc();
 }
 
+/// Determine whether the given "if" chain has a missing "else".
+static bool hasMissingElseInChain(IfStmt *ifStmt) {
+  if (!ifStmt->getElseStmt())
+    return true;
+
+  if (auto ifElse = dyn_cast<IfStmt>(ifStmt->getElseStmt()))
+    return hasMissingElseInChain(ifElse);
+
+  return false;
+}
+
 void SkipUnhandledConstructInFunctionBuilderFailure::diagnosePrimary(
     bool asNote) {
-  if (unhandled.is<Stmt *>()) {
+  if (auto stmt = unhandled.dyn_cast<Stmt *>()) {
     emitDiagnostic(asNote ? diag::note_function_builder_control_flow
                           : diag::function_builder_control_flow,
                    builder->getName());
+
+    // Emit custom notes to help the user introduce the appropriate 'build'
+    // functions.
+    SourceLoc buildInsertionLoc = builder->getBraces().Start;
+    std::string stubIndent;
+    Type componentType;
+    if (buildInsertionLoc.isValid()) {
+      buildInsertionLoc = Lexer::getLocForEndOfToken(
+          getASTContext().SourceMgr, buildInsertionLoc);
+
+      ASTContext &ctx = getASTContext();
+      StringRef extraIndent;
+      StringRef currentIndent = Lexer::getIndentationForLine(
+          ctx.SourceMgr, buildInsertionLoc, &extraIndent);
+      stubIndent = (currentIndent + extraIndent).str();
+
+      componentType = inferFunctionBuilderComponentType(builder);
+    }
+
+    if (buildInsertionLoc.isInvalid()) {
```

Comment:
I take it this will prevent us from showing these notes for declarations from the SDK. What about other modules in e.g. the same Xcode project? Will swiftsourceinfo files give us source locations for those?

Reply:
Yes, that's correct. `swiftsourceinfo` files should make it work with other modules in the same project/workspace.